### ISSUE TITLE
Transient and Request scopes do not work together

### ISSUE THREAD

**Issue by aliu-bsm** [nestjs/nest#3303]
## Current behavior
If the application has both `transient` and `request` scopes, the `transient` scoped service does not work. The services/controllers that use the `transient` service does not have the correct instance of the `transient` service

## Input Code
[Minimum reproducible repo](https://github.com/aliu-bsm/nestjs-scope)

code to run.
```bash
$ npm run start
...
$ curl http://localhost:3000
```
---

* `LoggerService` is a transient service
* `ContextService` is a request service

The `LoggerService` is an empty object when a request comes in. Everything initialized the `constructor` is gone. Services/Controllers are not using the same instance that was created. 

If either scope is removed, then both services work correctly according to its scope. It is only when there are transient and request scopes, then the transient scopes class won't work.

&nbsp;

Some code snippets: 
```ts
// logger.service.ts
@Injectable({
  scope: Scope.TRANSIENT,
})
export class LoggerService {
  prefix;
  logger;

  constructor() {
    this.logger = console;     // <--- ⚠️ It is called.
  }

  setPrefix(prefix: string) {
    this.prefix = prefix;
  }

  info(...args) {
    this.logger.log(`[${this.prefix}]`, ...args);
  }
}
```

```ts
// context.service.ts
@Injectable({
  scope: Scope.REQUEST,
})
export class ContextService {
  private context;

  setContext(obj: any): void {
    this.context = obj;
  }

  getContext(): any {
    return this.context;
  }
}
```

```ts
// app.controller.ts
@Controller()
export class AppController {
  constructor(
    private readonly log: LoggerService,
    private readonly context: ContextService,
    private readonly appService: AppService,
  ) {
    this.log.setPrefix('AppController');   // ⚠️ <-- `this.log.logger` is `undefined`
  }

  @Get()
  getHello(): string {
    this.context.setContext({
      time: new Date().toISOString(),
    });
    this.log.info('getHello');       // ⚠️ <-- `this.log.logger` is `undefined`
    return this.appService.getHello();
  }
}
```

```ts
// app.service.ts
@Injectable()
export class AppService {
  constructor(
    private readonly log: LoggerService,
    private readonly context: ContextService,
  ) {
    this.log.setPrefix('AppService');     // ⚠️ <-- `this.log.logger` is `undefined`
  }

  getHello(): string {
    this.log.info('getHello');      // ⚠️ <-- `this.log.logger` is `undefined`
    this.log.info(this.context.getContext());
    return 'Hello World!';
  }
}
```

## Expected behavior
The `LoggerService` instances should have used the one that was created instead of an empty object while in the request scope.

## Environment

<pre><code>
Nest version: 6.8.5 (latest at the moment)
 
For Tooling issues:
- Node version: `v10.16.3` and `v12.11.0` (both fails)
- Platform:  Mac

Others:
- NPM version: `6.9.0`
</code></pre>

**Comment by kamilmysliwiec** [comment]
Fixed in 6.9.0

**Comment by ulrichb** [comment]
Hi @kamilmysliwiec. Since 6.9.0 the following request-scoped service doesn't work anymore.

```ts
@Injectable({ scope: Scope.REQUEST })
export class ApiAuthenticatedUserProvider implements UserProvider {

    constructor(@Inject(REQUEST) private readonly request: express.Request) {
        console.log('!!!!!', { request });
    }

    get user(): ApiUser {
        return xxxxxxx(this.request.user);
    }
}
```

=> `request` is `null` and the c'tor is called way too early and only once (it seems to be interpreted as `SINGLETON`).

Registration happens via `{ provide: name, useClass: clazz }`.

**Comment by kamilmysliwiec** [comment]
@ulrichb if you use custom providers syntax (`provide:`) you should also define `scope` property: https://docs.nestjs.com/fundamentals/injection-scopes#usage (in this case, `scope: Scope.REQUEST`)

**Comment by ulrichb** [comment]
@kamilmysliwiec: Okay thanks, this fixes my problem. But why isn't it inferred by the decorator (as it seems to be in <= 6.8.5). Further it's a bit annoying to duplicate the scope information (also error prune when when doing refactorings).

**Comment by ulrichb** [comment]
Also note that I'm using `useClass` not `useValue` or `useFactory` (where it's clear that the `scope` must be specifed explicitly).

**Comment by kamilmysliwiec** [comment]
I'll push another fix shortly (to make it more straightforward :))

**Comment by ulrichb** [comment]
Okay cool. BTW: If not inferred by the decorator Nest should at least warn about a "mismatching" scope value when using `useClazz` custom providers a scope-value is present in the decorator.

**Comment by ulrichb** [comment]
@kamilmysliwiec Did you already have a look at this? Or should I create a new issue for my "`useClass` scope vs. decorator scope" issue?
